core/arch/arm/kernel/tee_time_arm_cntpct.c

```diff
@@ -45,15 +47,26 @@ REGISTER_TIME_SOURCE(arm_cntpct_time_source)
  *  - XORing it in 2-bit chunks with the whole CNTPCT contents
  *
  * and adding one byte of entropy when we reach 8 rotated bits.
+ *
+ * When enabled, we will attempt to collect SESSION jitter from the non-secure
+ * REE random number generator: an RPC call to the REE will add 196 bits
+ * of entropy to the pool. If the entropy data is not available, the algorithm
+ * fallbacks to using cntpct.
  */
 
 void plat_prng_add_jitter_entropy(enum crypto_rng_src sid, unsigned int *pnum)
 {
-	uint64_t tsc = barrier_read_cntpct();
+	uint64_t tsc = 0;
 	int bytes = 0, n;
 	static uint8_t first, bits;
 	static uint16_t acc;
 
+	if (IS_ENABLED(CFG_PRNG_SOURCE_RANDOM_REE) &&
+	    sid == CRYPTO_RNG_SRC_JITTER_SESSION &&
+	    !tee_random_add_ree_random(sid, pnum))
```

Comment:
This is far too often, it will add a significant overhead to RPCs. I'd say that this function should be called after `n` calls to `fortuna_read()` where `n` might be calculated with some random input. We shouldn't put too much trust in what normal world is feeding us. If normal world is owned by an attacker it has full control over these values. The jitter collected below might be a little harder to predict, at least the attacker will have to put some effort into predicting it accurately.

Reply:
ok.

but yes, without some way of verifying whether this helps or not probably makes no sense to merge anything on this area.

Reply:
OK